pkg/csi/service/fcd/controller.go

```diff
@@ -72,15 +75,123 @@ func (c *controller) CreateVolume(
+	zone := "TODO"
+	discoveryInfo, err := c.connMgr.WhichVCandDCByZone(ctx, zone)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to retrieve VC/DC based on zone %s. Err: %v", zone, err)
+	}
+
+	firstClassDisk, err := discoveryInfo.DataCenter.GetFirstClassDisk(
+		ctx, datastoreName, datastoreType, volName, vclib.FindFCDByName)
+	if err == nil {
+		log.Warningf("Volume with name %s already exists. Checking for similar parameters.", volName)
+
+		if firstClassDisk.Config.CapacityInMB != volSizeMB {
+			return nil, status.Errorf(codes.AlreadyExists,
+				"Volume already exists but requesting different size. Existing %d != Requested %d",
+				firstClassDisk.Config.CapacityInMB, volSizeMB)
+		}
+	} else {
+		err = discoveryInfo.DataCenter.CreateFirstClassDisk(ctx, datastoreName, datastoreType, volName, volSizeMB)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "CreateFirstClassDisk failed. Err: %v", err)
+		}
+
+		firstClassDisk, err = discoveryInfo.DataCenter.GetFirstClassDisk(
+			ctx, datastoreName, datastoreType, volName, vclib.FindFCDByName)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "GetFirstClassDiskByName(%s) failed. Err: %v", volName, err)
+		}
+	}
+
+	attributes := make(map[string]string)
+	attributes[AttributeFirstClassDiskType] = FirstClassDiskTypeString
+	attributes[AttributeFirstClassDiskName] = firstClassDisk.Config.Name
+	attributes[AttributeFirstClassDiskParentType] = string(firstClassDisk.ParentType)
+	if firstClassDisk.ParentType == vclib.TypeDatastoreCluster {
+		attributes[AttributeFirstClassDiskParentName] = firstClassDisk.StoragePodInfo.Summary.Name
+		attributes[AttributeFirstClassDiskOwningDatastore] = firstClassDisk.DatastoreInfo.Info.Name
+	} else {
+		attributes[AttributeFirstClassDiskParentName] = firstClassDisk.DatastoreInfo.Info.Name
+	}
+
+	resp := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			VolumeId:      firstClassDisk.Config.Id.Id,
+			CapacityBytes: int64(units.FileSize(firstClassDisk.Config.CapacityInMB * MbInBytes)),
+			VolumeContext: attributes,
+			//TODO: ContentSource?
+		},
+	}
+
+	return resp, nil
 }
 
 func (c *controller) DeleteVolume(
 	ctx context.Context,
 	req *csi.DeleteVolumeRequest) (
 	*csi.DeleteVolumeResponse, error) {
 
-	return nil, nil
+	//check for required parameters
+	if len(req.VolumeId) == 0 {
+		return nil, status.Errorf(codes.Internal, "Volume ID is a required parameter.")
+	}
+
+	discoveryInfo, err := c.connMgr.WhichVCandDCByFCDId(ctx, req.VolumeId)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to retrieve VC/DC based on FCDID %s. Err: %v", req.VolumeId, err)
+	}
+
+	// Volume Type
+	var datastoreName string
+	datastoreType := discoveryInfo.FCDInfo.ParentType
+	if datastoreType == vclib.TypeDatastore {
+		datastoreName = discoveryInfo.FCDInfo.DatastoreInfo.Info.Name
+	} else {
+		datastoreName = discoveryInfo.FCDInfo.StoragePodInfo.Summary.Name
+	}
+
```

Comment:
Should `DeleteVolume` be idempotent too and just return if the disk doesn't exist?

Reply:
Good call. If it doesn't exist, will just return success.